pincer/utils/tasks.py

```diff
@@ -0,0 +1,158 @@
+# Copyright Pincer 2021-Present
+# Full MIT License can be found in `LICENSE` at the project root.
+
+import asyncio
+import logging
+from asyncio import TimerHandle, iscoroutinefunction
+from datetime import timedelta
+from typing import Callable, Set
+
+
+from ..exceptions import (
+    TaskAlreadyRunning, TaskCancelError, TaskInvalidDelay,
+    TaskIsNotCoroutine
+)
+from . import __package__
+from .insertion import should_pass_cls
+from .types import Coro
+
+_log = logging.getLogger(__package__)
+
+
+class Task:
+    def __init__(self, scheduler: 'TaskScheduler', coro: Coro, delay: float):
+        """
+        A Task is a coroutine that is scheduled to repeat every x seconds.
+        Use a TaskScheduler in order to create a task.
+        """
+        self._scheduler = scheduler
+        self.coro = coro
+        self.delay = delay
+        self._handle: TimerHandle = None
+        self._client_required = should_pass_cls(coro)
+
+    def __del__(self):
+        if self.running:
+            self.cancel()
+        else:
+            # Did the user forgot to call task.start() ?
+            _log.warn(
+                "Task `%s` was not scheduled. Did you forget to start it ?",
+                self.coro.__name__
+            )
+
+    @property
+    def cancelled(self):
+        """Check if the task has been cancelled or not."""
+        return self.running and self._handle.cancelled()
+
+    @property
+    def running(self):
+        """Check if the task is running."""
+        return self._handle is not None
+
+    def start(self):
+        """Register the task in the TaskScheduler and start the execution of the task."""
```

Comment:
Make sure to respect the `pep8 & E501` for the docstring ^^
Seems hard but this is nothing crazy, just wrap the docstring and make the comments to have less than 72 characters, and have to code < 80 characters per line too.

If you are using pycharm, you can install the `.editorconfig` extension, which will allow you to load the config of the `.editorconfig` file with jb specific rules, that will show the 2 lines at 72 & 80 characters.

Reply:
Oops I'm sorry, I should have been more careful about it.
Thank you for the tip but I'm using vscode ^^ I'll add a ruler so I can spot the limit right away.

Reply:
no problem